### PR TITLE
Ports Whitesands13/Whitesands#766 "No More Chemspam: Redux"

### PR DESCRIPTION
**Pull request by rye-rice**
## About The Pull Request
Ports https://github.com/Whitesands13/Whitesands/pull/766

Here we have a perfectly fine and completed PR, ready to merge, yet it was never merged. Why? I have zero clue
Bascially ports TG botany, without the plant DNA manipulator removal. It also does not add the special tools meant to replace it

## Why It's Good For The Game

Good to keep the codebase updated with tg a bit, makes ghetto chem a lot more doable without stealing chem dispensers

The autogrow feature takes a shit ton of power, making more powerful power gen actually worth it

## Changelog
:cl: ArcaneMusic with tweaks from The0bserver
add: Adds and modifies fertilizers, as well as a new stat, Instability.
add: Upgrades plant guns to have two new functions. Revolution and Mutation Lock.
del: Removes the rather disused functionality of irrigation hoses.
tweak: Plant Analyzer have been upgraded. Using one in hand will now switch between a stat view of your plant, and a chemical view of your plant.
tweak: Trays now accept and store reagents, as well as coming with an autogrow mode. However, these upgrades came at the cost of old self sufficiency, meaning you must attend to your plants a bit more often.
tweak: Earthsblood, while not being able to gild trays anymore, has been found to still be quite powerful as a fertilizer.
/:cl:

### CODE REVIEW

File: code/modules/hydroponics/hydroitemdefines.dm [hunk 1]
```diff
@@ -10,7 +10,25 @@
...
+	if(ispodperson(M))
+		user.visible_message("<span class='notice'>[user] analyzes [M]'s vitals.</span>", \
+							"<span class='notice'>You analyze [M]'s vitals.</span>")
+		if(scan_mode== PLANT_SCANMODE_STATS)
+			healthscan(user, M, advanced = TRUE)
+		else
+			chemscan(user, M)
+		add_fingerprint(user)
+		return
+	return ..()
```
Comment (Fikou) [inline code/modules/hydroponics/hydroitemdefines.dm:31]: newline

File: code/modules/hydroponics/hydroponics_chemreact.dm [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
+ *This is NOW the gradual affects that each chemical applies on every process() proc. Nutrients now use a more robust reagent holder in order to apply less insane
+ * stat changes as opposed to 271 lines of individual statline effects. Shoutout to the original comments on chems, I just cleaned a few up.
+ */
+/obj/machinery/hydroponics/proc/apply_chemicals(mob/user)
+	///Contains the reagents within the tray.
+	if(myseed)
+		myseed.on_chem_reaction(reagents) //In case seeds have some special interactions with special chems, currently only used by vines
+	for(var/c in reagents.reagent_list)
+		var/datum/reagent/chem = c
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics_chemreact.dm:10]: as anything, descript variable names

File: code/modules/hydroponics/plant_genes.dm [hunk 3]
```diff
@@ -90,6 +90,14 @@
 	S.potency = value
 
 
+/datum/plant_gene/core/instability
+	name = "Stability"
+	value = 10
+
+/datum/plant_gene/core/instability/apply_stat(obj/item/seeds/S)
```
Comment (Fikou) [inline code/modules/hydroponics/plant_genes.dm:97]: descript variable name

File: code/modules/hydroponics/seeds.dm [hunk 4]
```diff
@@ -73,6 +77,10 @@
 /obj/item/seeds/examine(mob/user)
 	. = ..()
 	. += "<span class='notice'>Use a pen on it to rename it or change its description.</span>"
+	if(reagents_add && user.can_see_reagents())
+		. += "<span class='notice'>- Plant Reagents -</span>"
+		for(var/datum/plant_gene/reagent/G in genes)
+			. += "<span class='notice'>- [G.get_name()] -</span>"
```
Comment (Fikou) [inline code/modules/hydroponics/seeds.dm:83]: as anything, descript variable name

File: code/modules/reagents/chemistry/reagents.dm [hunk 5]
```diff
@@ -201,6 +201,15 @@ GLOBAL_LIST_INIT(name2reagent, build_name2reagent())
...
 	return
+/**
+  * New, standardized method for chemicals to affect hydroponics trays.
+  * Defined on a per-chem level as opposed to by the tray.
+  * Can affect plant's health, stats, or cause the plant to react in certain ways.
+  */
+/datum/reagent/proc/on_hydroponics_apply(obj/item/seeds/myseed, datum/reagents/chems, obj/machinery/hydroponics/mytray, mob/user)
+	if(!mytray || !chems)
+		return
+	return
```
Comment (Fikou) [inline code/modules/reagents/chemistry/reagents.dm:212]: what are you doing here

File: code/modules/hydroponics/seeds.dm [hunk 6]
```diff
@@ -253,6 +299,14 @@
 		if(C)
 			C.value = production
 
+/obj/item/seeds/proc/adjust_instability(adjustamt)
+	if(instability == -1)
+		return
+	instability = clamp(instability + adjustamt, 0, 100)
+	var/datum/plant_gene/core/C = get_gene(/datum/plant_gene/core/instability)
```
Comment (Fikou) [inline code/modules/hydroponics/seeds.dm:306]: yep

File: code/modules/hydroponics/seeds.dm [hunk 7]
```diff
@@ -310,6 +364,14 @@
 		if(C)
 			C.value = potency
 
+/obj/item/seeds/proc/set_instability(adjustamt)
+	if(instability == -1)
+		return
+	instability = clamp(adjustamt, 0, 100)
+	var/datum/plant_gene/core/C = get_gene(/datum/plant_gene/core/instability)
```
Comment (Fikou) [inline code/modules/hydroponics/seeds.dm:371]: yep

File: code/modules/hydroponics/seeds.dm [hunk 8]
```diff
@@ -310,6 +364,14 @@
 		if(C)
 			C.value = potency
 
+/obj/item/seeds/proc/set_instability(adjustamt)
+	if(instability == -1)
```
Comment (Fikou) [inline code/modules/hydroponics/seeds.dm:368]: this should be a define

**Review by Fikou — CHANGES_REQUESTED** [review]

File: code/modules/hydroponics/hydroponics.dm [hunk 9]
```diff
@@ -10,49 +6,64 @@
...
 	var/toxic = 0			//Toxicity in the tray?
 	var/age = 0				//Current age
-	var/dead = 0			//Is it dead?
+	var/dead = FALSE			//Is it dead?
 	var/plant_health		//Its health
 	var/lastproduce = 0		//Last time it was harvested
 	var/lastcycle = 0		//Used for timing of cycles.
-	var/cycledelay = CYCLE_DELAY_DEFAULT		// WS edit - Crystals
-	var/harvest = 0			//Ready to harvest?
+	var/cycledelay = CYCLE_DELAY_DEFAULT	//WS edit - Crystals
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:24]: kill the ws edit comments

File: code/modules/hydroponics/hydroponics.dm [hunk 10]
```diff
@@ -10,49 +6,64 @@
...
+	var/harvest = FALSE			//Ready to harvest?
 	var/obj/item/seeds/myseed = null	//The currently planted seed
 	var/rating = 1
-	var/unwrenchable = 1
+	var/unwrenchable = TRUE
 	var/recent_bee_visit = FALSE //Have we been visited by a bee recently, so bees dont overpollinate one plant
-	var/using_irrigation = FALSE //If the tray is connected to other trays via irrigation hoses
-	var/self_sufficiency_req = 20 //Required total dose to make a self-sufficient hydro tray. 1:1 with earthsblood.
-	var/self_sufficiency_progress = 0
+	var/mob/lastuser //Last user to add reagents to a tray. Mostly for logging.
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:30]: not a fan, can you weakref this or use some other method?

File: code/modules/hydroponics/hydroponics.dm [hunk 11]
```diff
@@ -10,49 +6,64 @@
...
 	var/self_sustaining = FALSE //If the tray generates nutrients and water on its own
-	var/mutate_yield = 0 // WS spritercode edit: makes L4Z worth a damn
+	// Here lies irrigation. You won't be missed, because you were never used.
 
+/obj/machinery/hydroponics/Initialize()
+	//Here lies "nutrilevel", killed by ArcaneMusic 20??-2019. Finally, we strive for a better future. Please use "reagents" instead
+	create_reagents(20)
+	reagents.add_reagent(/datum/reagent/plantnutriment/eznutriment, 10) //Half filled nutrient trays for dirt trays to have more to grow with in prison/lavaland.
+	. = ..()
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:38]: return ..() at end of procs

File: code/modules/hydroponics/hydroponics.dm [hunk 12]
```diff
@@ -68,59 +79,51 @@
...
 		return ..()
 	if(istype(Proj , /obj/projectile/energy/floramut))
 		mutate()
+		return BULLET_ACT_HIT
 	else if(istype(Proj , /obj/projectile/energy/florayield))
 		return myseed.bullet_act(Proj)
+	else if(istype(Proj , /obj/projectile/energy/florarevolution))
+		if(myseed)
+			if(myseed.mutatelist.len > 0)
+				myseed.instability = (myseed.instability/2)
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:98]:
```suggestion
		if(length(myseed.mutatelist))
			myseed.instability *= 0.5
```

File: code/modules/hydroponics/hydroponics.dm [hunk 13]
```diff
@@ -68,59 +79,51 @@
...
 	var/needs_update = 0 // Checks if the icon needs updating so we don't redraw empty trays every time
 	var/temp_sustain = FALSE	// If we want self_sustaining effects temporarily		// WS edit begin - Crystals
 
-	if(myseed)
-		if(myseed.loc != src)
-			myseed.forceMove(src)
-		if(myseed.get_gene(/datum/plant_gene/trait/plant_type/crystal))
-			temp_sustain = TRUE
+	if(myseed && (myseed.loc != src))
+		myseed.forceMove(src)
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:108]: dont run this on process, move the myseed on the exited proc

File: code/modules/hydroponics/hydroponics.dm [hunk 14]
```diff
@@ -396,10 +383,14 @@
 	harvest = 0
 	weedlevel = 0 // Reset
 	pestlevel = 0 // Reset
-	cycledelay = CYCLE_DELAY_DEFAULT		// WS edit - crystals
+	cycledelay = CYCLE_DELAY_DEFAULT //WS edit - crystals
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:386]: cut this

File: code/modules/hydroponics/hydroponics.dm [hunk 15]
```diff
@@ -821,41 +541,48 @@
...
+				to_chat(user, "*** <B>[myseed.plantname]</B> ***" )
+				to_chat(user, "- Plant Age: <span class='notice'>[age]</span>")
+				var/list/text_string = myseed.get_analyzer_text()
+				if(text_string)
+					to_chat(user, text_string)
+					to_chat(user, "*---------*")
+			if(myseed.reagents_add && P_analyzer.scan_mode == PLANT_SCANMODE_CHEMICALS)
+				to_chat(user, "- <B>Plant Reagents</B> -")
+				to_chat(user, "*---------*")
+				for(var/datum/plant_gene/reagent/G in myseed.genes)
```
Comment (Fikou) [inline code/modules/hydroponics/hydroponics.dm:570]: as anything, descript variable

**Comment by rye-rice** [comment]
See https://github.com/shiptest-ss13/Shiptest/pull/1116

**Pull request closed without merging**